### PR TITLE
i103 - Custom Data Stream Job

### LINKED ISSUE #103: Add custom DataStream Job type
We have a custom job type for Update and Cancel jobs.

We might as well add one for just the generic version of reading from a DataStream.

This would avoid the need to explicitly get the reader and lookup via index, we can just pass the instance directly like we do in the Update and Cancel jobs. 

See if `CancelComplete` is a good candidate for this as well.

### PR DESCRIPTION
**Pull request by jkeon**
Adds a new custom job type for the most common case of scheduling a job based on the result of some TaskData.

### What is the current behaviour?

Previously we were using an Anvil custom job type that wasn't aware of the Task system and that resulted in a job that looked like this:

```csharp
        [BurstCompile]
        private struct PathfindOnSuccessJob : IAnvilJobForDefer
        {
            // BEGIN SCHEDULING -----------------------------------------------------------------------------
            public static JobHandle Schedule(JobHandle dependsOn,
                                             DataStreamJobData<PathfindSuccess> jobData,
                                             DataStreamScheduleInfo<PathfindSuccess> scheduleInfo)
            {
                PathfindOnSuccessJob job = new PathfindOnSuccessJob(jobData);
                return job.ScheduleParallel(scheduleInfo.DeferredNativeArrayScheduleInfo,
                                            scheduleInfo.BatchSize,
                                            dependsOn);
            }
            // END SCHEDULING -------------------------------------------------------------------------------

            [WriteOnly] private DataStreamPendingWriter<FollowPathStart> m_FollowPathStartWriter;
            [ReadOnly] private readonly DataStreamActiveReader<PathfindSuccess> m_PathfindSuccessReader;

            public PathfindOnSuccessJob(DataStreamJobData<PathfindSuccess> jobData)
            {
                m_FollowPathStartWriter = jobData.GetDataStreamWriter<FollowPathStart>();
                m_PathfindSuccessReader = jobData.GetDataStreamReader<PathfindSuccess>();
            }

            public void InitForThread(int nativeThreadIndex)
            {
                m_FollowPathStartWriter.InitForThread(nativeThreadIndex);
            }

            public void Execute(int index)
            {
                PathfindSuccess pathfindSuccess = m_PathfindSuccessReader[index];
                m_FollowPathStartWriter.Add(new FollowPathStart(pathfindSuccess.Entity));
            }
        }
```

Nothing wrong with it, but we can do better.

### What is the new behaviour?

The new custom job type knows that you're going to be reading from Task data because that's what you scheduled on and it can optimize the flow allowing you to write less code. The same job as above now looks like this:

```csharp
        [BurstCompile]
        private struct PathfindOnSuccessJob : ITaskJobForDefer<PathfindSuccess>
        {
            // BEGIN SCHEDULING -----------------------------------------------------------------------------
            public static JobHandle Schedule(JobHandle dependsOn,
                                             DataStreamJobData<PathfindSuccess> jobData,
                                             DataStreamScheduleInfo<PathfindSuccess> scheduleInfo)
            {
                PathfindOnSuccessJob job = new PathfindOnSuccessJob(jobData);
                return job.ScheduleParallel(scheduleInfo,
                                            dependsOn);
            }
            // END SCHEDULING -------------------------------------------------------------------------------

            [WriteOnly] private DataStreamPendingWriter<FollowPathStart> m_FollowPathStartWriter;

            public PathfindOnSuccessJob(DataStreamJobData<PathfindSuccess> jobData)
            {
                m_FollowPathStartWriter = jobData.GetDataStreamWriter<FollowPathStart>();
            }

            public void InitForThread(int nativeThreadIndex)
            {
                m_FollowPathStartWriter.InitForThread(nativeThreadIndex);
            }

            public void Execute(PathfindSuccess pathfindSuccess)
            {
                m_FollowPathStartWriter.Add(new FollowPathStart(pathfindSuccess.Entity));
            }
        }
```

- Simpler scheduling
- No need to declare a reader explicitly and get it.
- No need to get the instance via index in the `Execute` function from the reader.
   - Side effect of better communicating intent and what you can do with it. You just get the readonly struct. With the index, you might think you can use that index to look up other related things which you can't do with the TaskSystem.

### What issues does this resolve?
- Resolves #103 

### What PRs does this depend on?
 - #143 

### Does this introduce a breaking change?
 - [ ] Yes
 - [x] No

### CODE REVIEW

File: Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskJobForDeferExtension.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                                              int jobIndex)
+            {
+                ref TJob jobData = ref wrapperData.JobData;
+                ref DataStreamActiveReader<TInstance> reader = ref wrapperData.Reader;
+                
+                jobData.InitForThread(wrapperData.NativeThreadIndex);
+
+                while (true)
+                {
+                    if (!JobsUtility.GetWorkStealingRange(ref ranges, jobIndex, out int beginIndex, out int endIndex))
```
Comment (mbaker3) [inline Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskJobForDeferExtension.cs:118]: Why not put this as the while conditional?

**Review by mbaker3 — COMMENTED** [review]
Nice! This looks really helpful.

Minor comments and one question about what looks like an unnecessary `ref`

File: Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/ITaskJobForDefer.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+using Anvil.Unity.DOTS.Data;
+using Unity.Jobs;
+using Unity.Jobs.LowLevel.Unsafe;
+
+namespace Anvil.Unity.DOTS.Entities.TaskDriver
+{
+    /// <summary>
+    /// A replacement for <see cref="IJobFor"/> when the number of work items is not known at Schedule time
+    /// and you are using a <see cref="DeferredNativeArray{T}"/>
+    /// This is specific to a context where the data is being read.
```
[the remark this answers is not in the document]
Comment (jkeon) [reply Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/ITaskJobForDefer.cs:10]: Sure, that makes sense. Fixed.

File: Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskJobForDeferExtension.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        // SCHEDULING
+        //*************************************************************************************************************
+
+        public static unsafe JobHandle ScheduleParallel<TJob, TInstance>(this TJob jobData,
+                                                                         DataStreamScheduleInfo<TInstance> scheduleInfo,
+                                                                         JobHandle dependsOn = default)
+            where TJob : struct, ITaskJobForDefer<TInstance>
+            where TInstance : unmanaged, IEntityProxyInstance
+        {
+            void* atomicSafetyHandlePtr = null;
```
[the remark this answers is not in the document]
Comment (jkeon) [reply Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskJobForDeferExtension.cs:22]: Hmmn, yeah this is a good idea. Changed for all job types that were doing it this way.

File: Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskJobForDeferExtension.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            atomicSafetyHandlePtr = scheduleInfo.DeferredNativeArrayScheduleInfo.SafetyHandlePtr;
+#endif
+
+            IntPtr reflectionData = WrapperJobProducer<TJob, TInstance>.JOB_REFLECTION_DATA;
+            ValidateReflectionData(reflectionData);
+            
+            WrapperJobStruct<TJob, TInstance> wrapperData = new WrapperJobStruct<TJob, TInstance>(ref jobData,
+                                                                                                  ref scheduleInfo);
```
[the remark this answers is not in the document]
Comment (jkeon) [reply Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskJobForDeferExtension.cs:32]: Yep, fixed.

File: Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskJobForDeferExtension.cs (hunk 1 above)
Comment (jkeon) [reply Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskJobForDeferExtension.cs:118]: I don't know, this was just how Unity had set up their job types and I followed suit when first making this. 

I just changed them in all our jobs and they all still work, so yeah, good catch.

**Review by mbaker3 — APPROVED** [review]

**Pull request merged**